src/dmd/backend/dlist.d

```diff
@@ -0,0 +1,571 @@
+    }
+    else
+    {
+        nlist++;
+        list = list_new();
+    }
+    return list;
+}
+
+list_t list_alloc(const(char)* file, int line)
+{
+    return list_alloc();
+}
+
+
+list_t list_new() { return cast(list_t)malloc(LIST.sizeof); }
+void list_delete(list_t list) { free(list); }
+
+/********************
+ * Free list.
+ * Params:
+ *      plist =         Pointer to list to free
+ *      freeptr =       Pointer to freeing function for the data pointer
+ *                      (use FPNULL if none)
+ * Output:
+ *      *plist is null
+ */
+
+void list_free(list_t* plist, list_free_fp freeptr)
+{
+    list_t list = *plist;
+    *plist = null;             // block any circular reference bugs
+    while (list && --list.count == 0)
+    {
+        list_t listnext = list_next(list);
+        if (freeptr)
+            (*freeptr)(list_ptr(list));
+        debug memset(list, 0, (*list).sizeof);
+        list.next = list_freelist;
+        list_freelist = list;
+        list = listnext;
+    }
+}
+
+void list_free(list_t *l)
+{
+     list_free(l, FPNULL);
+}
+
+/***************************
+ * Remove ptr from the list pointed to by *plist.
+ * Output:
+ *      *plist is updated to be the start of the new list
+ * Returns:
+ *      null if *plist is null
+ *      otherwise ptr
+ */
+
+void* list_subtract(list_t* plist, void* ptr)
+{
+    list_t list;
+
+    while ((list = *plist) != null)
+    {
+        if (list_ptr(list) == ptr)
+        {
+            if (--list.count == 0)
+            {
+                *plist = list_next(list);
+                list.next = list_freelist;
+                list_freelist = list;
+            }
+            return ptr;
+        }
+        else
+            plist = &list.next;
+    }
+    return null;            // it wasn't in the list
+}
+
+/***************************
+ * Remove first element in list pointed to by *plist.
+ * Returns:
+ *      First element, null if *plist is null
+ */
+
+void* list_pop(list_t* plist)
+{
+    return list_subtract(plist, list_ptr(*plist));
+}
+
+/*************************
+ * Append ptr to *plist.
+ * Returns:
+ *      pointer to list item created.
+ *      null if out of memory
+ */
+
+list_t list_append(list_t* plist, void* ptr)
+{
+    while (*plist)
+        plist = &(*plist).next;   // find end of list
+
+    list_t list = list_alloc();
+    if (list)
+    {
+        *plist = list;
+        list.next = null;
+        list.ptr = ptr;
+        list.count = 1;
+    }
+    return list;
+}
+
+list_t list_append_debug(list_t* plist, void* ptr, const(char)* file, int line)
+{
+    return list_append(plist, ptr);
+}
+
+/*************************
+ * Prepend ptr to *plist.
+ * Returns:
+ *      pointer to list item created (which is also the start of the list).
+ *      null if out of memory
+ */
+
+list_t list_prepend(list_t *plist, void *ptr)
+{
+    list_t list = list_alloc();
+    if (list)
+    {
+        list.next = *plist;
+        list.ptr = ptr;
```

Comment:
C implementation used the `list_next` and `list_ptr` wrappers.  It seems to be the same functionality, but throughout this file you're sometimes using the wrapper and sometimes not.  It makes it a little hard to review and the code is inconsistent.

Reply:
`list_next` and `list_ptr` don't work as lvalues in D.